### PR TITLE
Rolling upgrades: Migrate to ceph-key module

### PR DESCRIPTION
**Pull request by vishalkanaujia**
This change moves ceph-mgr upgrades to using ceph-key library.
Fixes: #2758

Signed-off-by: Vishal Kanaujia <vishal.kanaujia@flipkart.com>

### CODE REVIEW

**Comment by guits** [comment]
jenkins test luminous-ansible2.4-update_cluster

**Comment by guits** [comment]
jenkins test luminous-ansible2.4-update_docker_cluster

**Comment by leseb** [comment]
Please rebase on master.

**Comment by vishalkanaujia** [comment]
@leseb I don't know how it interpreted the rebase as "@vishalkanaujia dismissed leseb’s stale review via 193394e 2 minutes ago"

I have rebased the PR, request you to take a look.

**Comment by guits** [comment]
jenkins test pipeline

**Comment by leseb** [comment]
jenkins test luminous-ansible2.4-update_cluster

**Comment by leseb** [comment]
jenkins test luminous-ansible2.4-update_docker_cluster

**Comment by vishalkanaujia** [comment]
@leseb the second failure is due to docker config. I am trying to understand the cause of the first failure.

**Comment by guits** [comment]
@vishalkanaujia at first glance, the failure is unrelated to your patch. The data structure for cluster status (json output) in mimic has slightly changed.

**Comment by vishalkanaujia** [comment]
@guits thanks for the check. I have rebased the PR with master.

**Comment by guits** [comment]
jenkins test luminous-ansible2.4-update_cluster

**Comment by guits** [comment]
jenkins test luminous-ansible2.4-update_docker_cluster

**Comment by guits** [comment]
jenkins test luminous-ansible2.4-update_docker_cluster

**Comment by guits** [comment]
jenkins test luminous-ansible2.4-update_cluster

**Comment by guits** [comment]
@vishalkanaujia sorry, I've taken the liberty of rebasing your branch on master, which should fix your PR (thanks to a recent commit in master). Do not forget to `git reset --hard origin/fix_rolling_upgrade` on your local repo

**Comment by vishalkanaujia** [comment]
@guits The failure in second test is due to the following:

`The conditional check 'hostvars[mon_host]['ansible_hostname'] in (ceph_health_raw.stdout | from_json)["quorum_names"] or hostvars[mon_host]['ansible_fqdn'] in (ceph_health_raw.stdout | from_json)["quorum_names"]
' failed. The error was: No JSON object could be decoded`

I'm unable to get error logs for ceph-ansible PR pipeline test.

**Comment by vishalkanaujia** [comment]
@guits @leseb Please advise on the failing tests.

**Comment by leseb** [comment]
@vishalkanaujia please rebase on master. Thanks.

**Comment by vishalkanaujia** [comment]
@leseb I rebase the PR on the master.

File: infrastructure-playbooks/rolling_update.yml [hunk 1]
```diff
@@ -232,29 +232,42 @@
...
-
-    - name: container | create ceph mgr keyring(s)
-      command: "docker exec ceph-mon-{{ hostvars[groups[mon_group_name][0]]['ansible_hostname'] }} ceph --cluster {{ cluster }} auth get-or-create mgr.{{ hostvars[item]['ansible_hostname'] }} mon 'allow profile mgr' osd 'allow *' mds 'allow *' -o /etc/ceph/{{ cluster }}.mgr.{{ hostvars[item]['ansible_hostname'] }}.keyring"
-      args:
-        creates: "{{ ceph_conf_key_directory }}/{{ cluster }}.mgr.{{ hostvars[item]['ansible_hostname'] }}.keyring"
-      changed_when: false
+        - cephx
+        - groups.get(mgr_group_name, []) | length > 0
+        - ceph_release_num[ceph_release] >= ceph_release_num.luminous
+        - not containerized_deployment
```
Comment (leseb) [inline infrastructure-playbooks/rolling_update.yml]: you have `not containerized_deployment` twice

Comment (vishalkanaujia) [reply infrastructure-playbooks/rolling_update.yml]: Thanks, I deleted the line. Revised the PR.

File: infrastructure-playbooks/rolling_update.yml [hunk 2]
```diff
@@ -206,7 +206,7 @@
 
   vars:
     upgrade_ceph_packages: True
-
+    ceph_release: "{{ ceph_stable_release }}"
```
Comment (leseb) [inline infrastructure-playbooks/rolling_update.yml]: This does not seem to be connected to your commit, why this addition?

**Comment by vishalkanaujia** [comment]
@leseb: thanks again for the review. I removed the unwanted code and rebased the PR.

**Comment by leseb** [comment]
@vishalkanaujia thanks, waiting for the CI now.

**Comment by vishalkanaujia** [comment]
@leseb thanks you 👍

**Comment by vishalkanaujia** [comment]
@leseb  The test failed for the below reason:

`The conditional check 'ceph_release_num[ceph_release] >= ceph_release_num.luminous' failed. The error was: error while evaluating conditional (ceph_release_num[ceph_release] >= ceph_release_num.luminous): 'ceph_release' is undefined`

I remember that's why I had added the ceph_release variable under the upgrade ceph mgr node task. The variable `ceph_release` is set in the `ceph-defaults` role (`./roles/ceph-defaults/tasks/facts.yml`)

What else is a better way to get this information?

**Comment by leseb** [comment]
@vishalkanaujia the way you put it before was fine, please put it back, thanks

**Comment by vishalkanaujia** [comment]
@leseb  Thanks, I made the changes, rebased the PR with master.

**Comment by vishalkanaujia** [comment]
@leseb One test `ceph-ansible-prs-luminous-ansible2.4-update_docker_cluster` failed for the following reason:
```
TASK [get yum check-update result] *********************************************
task path: /home/jenkins-build/build/workspace/ceph-ansible-prs-luminous-ansible2.4-update_docker_cluster/collect-logs.yml:14
ok: [osd0] => {
    "changed": false, 
    "cmd": "yum check-update", 
    "failed_when_result": false, 
    "rc": 2
}

MSG:

[Errno 2] No such file or directory
ok: [mon2] => {
    "changed": false, 
    "cmd": "yum check-update", 
    "failed_when_result": false, 
    "rc": 2
}
```

**Comment by leseb** [comment]
@vishalkanaujia No the error is on testinfra. Not related to your PR.

**Comment by vishalkanaujia** [comment]
@leseb  thanks again. I'm rebasing the PR on the master.

**Comment by vishalkanaujia** [comment]
@leseb: requesting you to merge the PR.

**Comment by vishalkanaujia** [comment]
@leseb  The new failure appears in PR syntax check:

```
++ grep -c Signed-off-by
+ test 1 -ne 1
+ return 1
Build step 'Execute shell' marked build as failure
Finished: FAILURE
```

I could not find the problem with the PR syntax though.

**Comment by leseb** [comment]
@vishalkanaujia it's on our side

**Comment by leseb** [comment]
jenkins test pipeline

**Comment by leseb** [comment]
jenkins test pipeline

**Comment by guits** [comment]
unrelated failure, let's merge this as it is.

**Comment by vishalkanaujia** [comment]
@guits @leseb  Thanks so much.

**Pull request merged**